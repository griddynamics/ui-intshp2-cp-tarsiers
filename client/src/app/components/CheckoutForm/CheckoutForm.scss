@import '../../../styles/variables';
@import '../../../styles/mixin';

.checkout-form {
  color: inherit;
  text-align: start;
  letter-spacing: 3px;
  font-size: 16px;
  text-transform: uppercase;

  form {
    margin: 0 auto;
    display: block;
    width: 450px;
    font-size: 14px;
  }

  label {
    color: inherit;
    display: block;
    margin: 10px 0;
    position: relative;
    width: 100%;
  }

  fieldset {
    text-align: center;
    white-space: nowrap;
    padding: 10px;
    margin: 0 auto;
    width: 410px;
  }

  input {
    padding: 0 10px;
    margin: 0 auto;
    transition: 0.3s;
    font-size: 16px;
    outline: none;
    background-color: $color-light-1;
    border: none;
    border-bottom: 2px solid $color-main;
    display: block;
    width: 100%;
    height: 26px;
  }

  .success {
    border-bottom: 2px solid $color-success;
  }

  .tooltip {
    border-radius: 4px;
    padding: 2px;
    visibility: hidden;
    opacity: 0;
    position: absolute;
    display: block;
    width: 100%;
    top: 110%;
    height: 42px;
    background-color: $color-white;
    z-index: 1;
    text-align: middle;
    box-shadow: 0px 0px 10px $color-grey-1;
    transition: 0.3s;

    &::after {
      content: '';
      position: absolute;
      top: -10%;
      left: 10%;
      width: 0;
      height: 0;
      border-left: 5px solid transparent;
      border-right: 5px solid transparent;
      border-bottom: 5px solid $color-white;
    }
  }

  .show {
    visibility: visible;
    opacity: 1;
  }

  input:-internal-autofill-selected {
    background-color: $color-success;
  }

  .error {
    background-color: rgba($color-main, 0.5);
    border-bottom: 2px solid $color-alert;
  }
}
