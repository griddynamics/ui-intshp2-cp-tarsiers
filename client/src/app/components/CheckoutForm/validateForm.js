const patterns = {
  name: /^([a-zA-Z0-9]+|[a-zA-Z0-9]+\s{1}[a-zA-Z0-9]{1,}|[a-zA-Z0-9]+\s{1}[a-zA-Z0-9]{3,}\s{1}[a-zA-Z0-9]{1,}){6,}$/,
  email: /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/,
  address: /^.{6,}/,
  phoneNumber: /\+\(380\){1}[0-9]{9}$/
};

function validateForm() {
  const inputNodes = [...document.querySelectorAll('input')];

  inputNodes.forEach(el => {
    if (!this.validateField(el, el.id)) {
      el.classList.add('error');
    }
  });
  return !inputNodes.some(el => el.classList.contains('error'));
}

function validateField(node, field) {
  return patterns[field].test(node.value);
}

export { validateForm, validateField };
