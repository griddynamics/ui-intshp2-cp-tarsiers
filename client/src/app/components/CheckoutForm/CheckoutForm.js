/* eslint-disable no-param-reassign */
/* eslint-disable no-unused-vars */
import React, { Component } from 'react';
import IMask from 'imask';

import { validateForm, validateField } from './validateForm';
import StyledButton from '../../common/StyledButton';
import ls from '../../../utils/localStorage.service';

import './CheckoutForm.scss';

const inputPairs = {
  name: 'Full Name',
  email: 'Email',
  address: 'Address',
  phoneNumber: 'Phone Number'
};

const hints = {
  name: 'Full Name should be at least 6 characters long',
  email: 'Please enter valid email',
  address: 'Address should be at least 6 characters long',
  phoneNumber: 'Phone number should match this pattern "+(380)*********"'
};

class CheckoutForm extends Component {
  formFields = Object.entries(inputPairs).map(el => (
    <label htmlFor={el[0]} key={el[0]}>
      {el[1]}
      <input
        type="text"
        id={el[0]}
        onChange={e => this.handleInputChange(e, el[0])}
        onBlur={e => this.handleInputBlur(e, el[0])}
        onFocus={this.handleInputFocus}
      />
      <div className="tooltip" id={`${el[0]}-tooltip`}>
        {hints[el[0]]}
      </div>
    </label>
  ));

  constructor(props) {
    super(props);
    this.validateForm = validateForm.bind(this);
    this.validateField = validateField.bind(this);
    this.formRef = React.createRef();
  }

  componentDidMount() {
    const element = document.getElementById('phoneNumber');
    const maskOptions = { mask: '+({38\\0})000000000' };
    const mask = new IMask(element, maskOptions);

    this.findAllInputs();
  }

  findAllInputs = () => {
    const formData = ls.getState('form');

    if (formData) {
      const form = this.formRef.current;

      [...form.elements].forEach(el => {
        if (el.nodeName.toLowerCase() === 'input') {
          el.value = formData[el.id];
          if (this.validateField(el, el.id)) {
            el.classList.add('success');
          }
        }
      });
    }
  };

  handleInputBlur = (e, field) => {
    this.hideTooltip(e);
    if (!this.validateField(e.target, field)) {
      this.addErrorClass(e);
    }
  };

  handleInputFocus = e => {
    if (!e.currentTarget.classList.contains('success')) {
      this.showTooltip(e);
    }
    this.removeErrorClass(e);
  };

  hideTooltip = e => {
    e.currentTarget.nextSibling.classList.remove('show');
  };

  showTooltip = e => {
    e.currentTarget.nextSibling.classList.add('show');
  };

  addClassSuccess = e => {
    e.currentTarget.classList.add('success');
  };

  handleInputChange = (e, field) => {
    this.setState({ [field]: { value: e.target.value } });

    if (this.validateField(e.currentTarget, field)) {
      e.currentTarget.classList.add('success');
      this.hideTooltip(e);
    } else {
      e.currentTarget.classList.remove('success');
      this.showTooltip(e);
    }
  };

  removeErrorClass = e => {
    e.currentTarget.classList.remove('error');
  };

  addErrorClass = e => {
    e.currentTarget.classList.add('error');
  };

  onAcceptAction = e => {
    const { handleSubmit } = this.props;

    e.preventDefault();

    if (this.validateForm()) {
      const form = this.formRef.current;
      const data = {};

      [...form.elements].forEach(el => {
        if (el.nodeName.toLowerCase() === 'input') {
          data[el.getAttribute('id')] = el.value;
        }
      });

      ls.setState('form', data);
    }

    handleSubmit(this.validateForm);
  };

  onCancelAction = e => {
    e.preventDefault();
    const form = this.formRef.current;

    [...form.elements].forEach(el => {
      if (el.nodeName.toLowerCase() === 'input') {
        el.value = '';
      }
    });

    ls.setState('form');
    this.validateForm();
  };

  render() {
    return (
      <div className="container checkout-form">
        <form ref={this.formRef} autoComplete="off">
          {this.formFields}
          <fieldset>
            <StyledButton
              type="submit"
              btnClass="accept"
              callback={this.onAcceptAction}
            />
            <StyledButton
              type="submit"
              btnClass="cancel"
              callback={this.onCancelAction}
            />
          </fieldset>
        </form>
      </div>
    );
  }
}

export default CheckoutForm;
