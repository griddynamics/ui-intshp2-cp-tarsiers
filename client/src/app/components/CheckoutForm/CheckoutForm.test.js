import React from 'react';
import { shallow, mount } from 'enzyme';
import CheckoutForm from './CheckoutForm';

jest.mock('imask');
xdescribe('<CheckoutForm />', () => {
  it('should match its snapshot', () => {
    const wrapper = shallow(<CheckoutForm />);

    expect(wrapper).toMatchSnapshot();
  });

  it('component reacts properly on input change', () => {
    const wrapper = mount(<CheckoutForm />);
    const value = '80% of developers earn 20% of money';
    const evt = { target: { value } };

    wrapper.find('#name').simulate('change', evt);

    expect(wrapper.state().name.value).toBe(value);
  });

  it('component should call all the functions when necessary', () => {
    const wrapper = mount(<CheckoutForm />);

    jest.spyOn(wrapper.instance(), 'handleInputBlur');
    jest.spyOn(wrapper.instance(), 'handleInputFocus');

    wrapper.find('#name').simulate('focus');
    wrapper.find('#name').simulate('blur');

    expect(wrapper.instance().handleInputBlur).toHaveBeenCalled();
    expect(wrapper.instance().handleInputFocus).toHaveBeenCalled();
  });
});
