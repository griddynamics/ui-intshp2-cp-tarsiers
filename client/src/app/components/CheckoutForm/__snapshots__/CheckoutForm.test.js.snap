// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutForm /> should match its snapshot 1`] = `
<div
  className="container checkout-form"
>
  <form
    autoComplete="off"
    onSubmit={[Function]}
  >
    <label
      htmlFor="name"
      key="name"
    >
      Full Name
      <input
        id="name"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
      />
      <div
        className="tooltip"
        id="name-tooltip"
      >
        Full Name should be at least 6 characters long
      </div>
    </label>
    <label
      htmlFor="email"
      key="email"
    >
      Email
      <input
        id="email"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
      />
      <div
        className="tooltip"
        id="email-tooltip"
      >
        Please enter valid email
      </div>
    </label>
    <label
      htmlFor="address"
      key="address"
    >
      Address
      <input
        id="address"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
      />
      <div
        className="tooltip"
        id="address-tooltip"
      >
        Address should be at least 6 characters long
      </div>
    </label>
    <label
      htmlFor="phoneNumber"
      key="phoneNumber"
    >
      Phone Number
      <input
        id="phoneNumber"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        type="text"
      />
      <div
        className="tooltip"
        id="phoneNumber-tooltip"
      >
        Phone number should match this pattern "+(380)*********"
      </div>
    </label>
    <button
      type="submit"
    >
      Proceed
    </button>
  </form>
</div>
`;
